clippy_lints/src/set_contains_or_insert.rs

```diff
@@ -112,6 +114,45 @@ fn try_parse_op_call<'tcx>(
     None
 }
 
+fn is_set_mutated<'tcx>(cx: &LateContext<'tcx>, contains_expr: &OpExpr<'tcx>, expr: &'tcx Expr<'_>) -> bool {
+    let expr = peel_hir_expr_while(expr, |e| {
+        if let ExprKind::Unary(UnOp::Not, e) = e.kind {
+            Some(e)
+        } else {
+            None
+        }
+    });
+
+    if let ExprKind::MethodCall(_, receiver, ..) = expr.kind
+        && let receiver = receiver.peel_borrows()
+        && let receiver_ty = cx.typeck_results().expr_ty(receiver).peel_refs()
+        && matches!(receiver_ty.opt_diag_name(cx), Some(sym::HashSet | sym::BTreeSet))
+        && SpanlessEq::new(cx).eq_expr(contains_expr.receiver, receiver)
+        && let Some(method_def) = cx.typeck_results().type_dependent_def_id(expr.hir_id)
+        && let method_fn_sig = cx.tcx.fn_sig(method_def).instantiate_identity().skip_binder()
+        && let Some(self_ty) = method_fn_sig.inputs().first()
+        && let ty::Ref(_, _, ty::Mutability::Mut) = self_ty.kind()
+    {
+        return true;
+    }
+
+    let child = expr.peel_borrows();
+    let child_ty = cx.typeck_results().expr_ty(child).peel_refs();
+    if matches!(child_ty.opt_diag_name(cx), Some(sym::HashSet | sym::BTreeSet))
+        && SpanlessEq::new(cx).eq_expr(contains_expr.receiver, child)
+        && let Node::Expr(parent) = cx.tcx.parent_hir_node(expr.hir_id)
+        && let ExprKind::Call(func, args) = parent.kind
+        && let Some(func_sig) = expr_sig(cx, func)
+        && let Some(set_index) = args.iter().position(|arg| arg.hir_id == expr.hir_id)
+        && let Some(set_ty) = func_sig.input(set_index).map(Binder::skip_binder)
+        && let ty::Ref(_, _, ty::Mutability::Mut) = set_ty.kind()
+    {
+        return true;
+    }
+
+    false
```

Comment:
I think you can look for usages of the receiver in a context where it has not been adjusted to a non-mutable pointer:

```suggestion
    // Guard on type to avoid useless potentially expansive `SpanlessEq` checks
    cx.typeck_results().expr_ty_adjusted(expr).is_mutable_ptr()
        && matches!(
            cx.typeck_results().expr_ty(expr).peel_refs().opt_diag_name(cx),
            Some(sym::HashSet | sym::BTreeSet)
        )
        && SpanlessEq::new(cx).eq_expr(contains_expr.receiver, expr.peel_borrows())
```

It will miss the case where you explicitly reborrow, for example to do `(&*set).contains(&something_else)` but such code would probably not exist in real life, and a false negative in such a peculiar situation would not be a big deal.

Can you add this test?

```rust
    if !s.contains(&v) {
        //~^ set_contains_or_insert
        let _readonly_access = s.contains(&v);
        s.insert(v);
    }
```

It explicitly checks that an implicit reborrow inserted by the compiler is enough not to prevent the lint from triggering.